Enabling XML format for UK

jurgenbosch opened the pull request.

monofone commented:
HI @jurgenbosch , this looks pretty neat. Can you tell be what the docker-compose and Dockerfile are used for ? I assume for testing but I didn't get the connection how this is build and run.

jurgenbosch commented:
@monofone Thank you for your comment. I added these files for fast rapid local development and wanted to remove the local dependencies. I removed them from this PR for the sake of keeping things clean. 

If you agree with me that Docker would be an addition to the repo I will re-add the files in a separate PR.

monofone commented:
Hi @jurgenbosch, thanks for the changes, I would appreciate a good development setup, for this I'll be happy to see an additional PR. Would be great to have a development section in the Readme for this, explaining how to use the setup.

monofone commented on src/PaymentInformation.php:
```diff
@@ -164,6 +164,21 @@ class PaymentInformation
+    /**
+     * @var bool
+     */
+    protected $showCurrency = true;
```
Se comment above for `&& $paymentInformation->showCurrency()`

jurgenbosch replied:
See comment above. This is for backwards compatibility.

monofone replied:
Now I am back on track, this is not the suitable place as this is about how to build the XML files, this should reside in the DOMBuilders and not in the data structure.

monofone reviewed (changes requested):
Hi @jurgenbosch I've taken a second look, beside it is still looking good with the tests an everything I found two issues I like to discuss.

jurgenbosch commented on src/DomBuilder/CustomerCreditTransferDomBuilder.php:
```diff
@@ -93,6 +93,15 @@ public function visitPaymentInformation(PaymentInformation $paymentInformation):
         $debtor->appendChild($this->createElement('Nm', $paymentInformation->getOriginName()));
+
+
+        if ($paymentInformation->getOriginAccountCurrency() === 'GBP') {
```
So the pain format is equal for both EU & UK payments. 
For the UK market the output format differs in some places. Examples of this are;
- The need of a PstlAdr object with TwnNm & Ctry fields 
- Obviously have a different currency in the InstdAmt 
- There is a limit of 16 chars on the EndToEndId field
- The field BIC is mandatory in the CdtTrfTxInf object
- The ServiceLevel code needs to be NURG instead of SEPA
- The schemaLocation needed to be removed from the output 
- The ChrgBr field needed to be removed. 

All these requirements above made me implement them in the best way possible. I didn't want to introduce a new property in order to be fully backwards compatible.

jurgenbosch commented on src/DomBuilder/CustomerCreditTransferDomBuilder.php:
```diff
@@ -107,7 +116,7 @@ public function visitPaymentInformation(PaymentInformation $paymentInformation):
         $id->appendChild($this->createElement('IBAN', $paymentInformation->getOriginAccountIBAN()));
         $debtorAccount->appendChild($id);
-        if ($paymentInformation->getOriginAccountCurrency()) {
+        if ($paymentInformation->getOriginAccountCurrency() && $paymentInformation->showCurrency()) {
```
This statement is modified to provide control if the implementation want to show or hide the currency (Ccy) field in the DebtorAccount information. As you mentioned it defaults to true and is never changed in the package but it is changed in my implementation as the field is required not to be there in the UK.

monofone commented on tests/Unit/DomBuilder/CustomerCreditTransferDomBuilderTest.php:
```diff
@@ -0,0 +1,93 @@
+        $this->assertSame('NURG', $xpath->evaluate('/ns:Document/ns:CstmrCdtTrfInitn/ns:PmtInf/ns:PmtTpInf/ns:SvcLvl/ns:Cd')->item(0)->textContent);
+        $this->assertNull($xpath->evaluate('/ns:Document/ns:CstmrCdtTrfInitn/ns:PmtInf/ns:ChrgBr')->item(0));
+        $this->assertSame('For my lovely colleague, Nerd', $xpath->evaluate('/ns:Document/ns:CstmrCdtTrfInitn/ns:PmtInf/ns:CdtTrfTxInf/ns:RmtInf/ns:Ustrd')->item(0)->textContent);
+        $this->assertSame('Ccy', $xpath->evaluate('/ns:Document/ns:CstmrCdtTrfInitn/ns:PmtInf/ns:CdtTrfTxInf/ns:Amt/ns:InstdAmt')->item(0)->attributes->item(0)->name);
```
Shouldn't this attribute be absent ? as I understood your comment correctly 

> This statement is modified to provide control if the implementation want to show or hide the currency (Ccy) field in the DebtorAccount information. As you mentioned it defaults to true and is never changed in the package but it is changed in my implementation as the field is required not to be there in the UK.

monofone reviewed (changes requested):
Hi again,

now I got some time to take a deeper look. Most of the chnages are fine for me. But can we remove al implementation specific stuff regarding the UK case ? Then I am ok to merge this.
I added an additional example on how to get your special case done. 

An Idea would be to add an additional repository with special DomBuilder Implementations for special cases like yours.

For now I just added a `CustomerCreditTransferDomBuilderUk` to the repository but would suggest that you implement this one in your application code and not in the library as this would lead to a lot of different DomBuilders spamming the library.

 Beside this I wish you a happy and healthy new year.

Regards

allphat commented:
I had to implement transfer from UK to UK customers. It was a real pain. I had to override some parts of the original code (for example, I had to use BBAN instead of IBAN/BIC).
I  think the bank I am using has its own format based on the pain format.
I just hope most of UK banks are not the same :)

Just to warn that not all uk banks will accept this format

BorislavSabev commented:
Closing this PR as it is a specific and 'out of scope' case.
